sema/subyt/sources.py

```diff
@@ -0,0 +1,333 @@
+            % self._collection_path
+        )
+        self._reset()
+        self.mtimes = {}
+        for p in self._sourcefiles:
+            p = Path(self._collection_path) / Path(p)
+            self.mtimes.update({str(p): os.stat(p).st_mtime})
+
+
+try:
+    import csv
+
+    class CSVFileSource(Source):
+        """
+        Source producing iterator over data-set coming from CSV on file
+        """
+
+        def __init__(self, csv_file_path):
+            super().__init__()
+            assert_readable(csv_file_path)
+            self._csv = csv_file_path
+            if Path(csv_file_path).exists():
+                self.mtimes = {csv_file_path: os.stat(csv_file_path).st_mtime}
+
+        def __enter__(self):
+            self._csvfile = open(self._csv, mode="r", encoding="utf-8-sig")
+            return csv.DictReader(self._csvfile, delimiter=",")
+
+        def __exit__(self):
+            self._csvfile.close()
+
+        def __repr__(self):
+            return "CSVFileSource('%s')" % os.path.abspath(self._csv)
+
+    SourceFactory.register("text/csv", CSVFileSource)
+    # wrong, yet useful mime for csv:
+    SourceFactory.register("application/csv", CSVFileSource)
+except ImportError:
+    log.warn("Python CSV module not available -- disabling CSV support!")
+
+
+try:
+    import json
+
+    class JsonFileSource(Source):
+        """
+        Source producing iterator over data-set coming from json on file
+        """
+
+        def __init__(self, json_file_path):
+            super().__init__()
+            assert_readable(json_file_path)
+            self._json = json_file_path
+            if Path(json_file_path).exists():
+                self.mtimes = {
+                    json_file_path: os.stat(json_file_path).st_mtime
+                }
+
+        def __enter__(self):
+            # note this is loading everything in memory
+            #   -- will not work for large sets
+            #   in the future we might need to consider:
+            #   (json-stream)[https://pypi.org/project/json-stream/]
+            with open(self._json, mode="r", encoding="utf-8-sig") as jsonfile:
+                data = json.load(jsonfile)
+                # unwrap nested structures
+                while isinstance(data, dict) and len(data.keys()) == 1:
+                    child = list(data.values())[0]
+                    if isinstance(child, dict) and len(child.keys()) == 0:
+                        data = [data]
+                    else:
+                        data = child
+                if not isinstance(data, list):
+                    data = [data]
+            return iter(data)
+
+        def __exit__(self):
+            pass
+
+        def __repr__(self):
+            return "JsonFileSource('%s')" % os.path.abspath(self._json)
+
+    SourceFactory.register("application/json", JsonFileSource)
+except ImportError:
+    log.warn("Python JSON module not available -- disabling JSON support!")
+
+
+try:
+    import xmlasdict
```

Comment:
this ensures the xmlasdict is not a hard requirement --> could be moved to test or dev group !